### PR TITLE
PPTP-1243 - initial integration with address lookup frontend

### PR DESCRIPTION
**Pull request by SteveSugden**
Integrate with address lookup frontend to collect contact address

Requires 
Service manager - https://github.com/hmrc/service-manager-config/pull/3956

#### Check list 
 - [x] `./precheck` was executed (Integration/Component/Unit tests)
 - [x] `sbt scalafmt test:scalafmt` was executed
 - [ ] Required Environment Config has been amended/added
 - [ ] Links to dependencies have been included (BE/FE work)
 - [ ] User Acceptance Tests (UAT) were run locally and have passed.
 - [ ] No Accessibility errors/warnings reported by Wave

### CODE REVIEW

File: app/uk/gov/hmrc/plasticpackagingtax/registration/models/addresslookup/AddressLookupConfigV2.scala [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+case class AddressLookupConfigV2(version: Int = 2, options: JourneyOptions, labels: JourneyLabels)
+
+object AddressLookupConfigV2 {
+  implicit val format: OFormat[AddressLookupConfigV2] = Json.format[AddressLookupConfigV2]
+
+  def apply(continue: Call, appConfig: AppConfig)(implicit
+    messagesApi: MessagesApi,
+    request: JourneyRequest[_]
+  ): AddressLookupConfigV2 = {
+    val en: Messages = MessagesImpl(Lang("en"), messagesApi)
```
[the remark this answers is not in the document]
Comment (RyanKellyHMRC) [reply app/uk/gov/hmrc/plasticpackagingtax/registration/models/addresslookup/AddressLookupConfigV2.scala:36]: Or would we specifically enable Welsh support later?

File: app/uk/gov/hmrc/plasticpackagingtax/registration/models/addresslookup/AddressLookupConfirmation.scala [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+  // modified from the address-lookup-frontend, when it wants to split the address for manual edit
+  def extractAddressLines(): (String, Option[String], Option[String], String) = {
+    val l1: String         = this.address.lines.headOption.getOrElse("")
+    val l2: Option[String] = if (this.address.lines.length > 2) this.address.lines.lift(1) else None
+    val l3: Option[String] = if (this.address.lines.length > 3) this.address.lines.lift(2) else None
+    val l4: String =
+      if (this.address.lines.length > 1) this.address.lines.lastOption.getOrElse("") else ""
+    (l1, l2, l3, l4)
+  }
```
Comment (RyanKellyHMRC) [inline app/uk/gov/hmrc/plasticpackagingtax/registration/models/addresslookup/AddressLookupConfirmation.scala:35]: Might be worth adding some explicit tests for this?

**Review by RyanKellyHMRC — APPROVED** [review]
I'm happy with it. A couple of questions and suggestion that some explicit tests might be worth adding.

File: app/uk/gov/hmrc/plasticpackagingtax/registration/connectors/addresslookup/AddressLookupFrontendConnector.scala [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    addressLookupRequest: AddressLookupConfigV2
+  )(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[AddressLookupOnRamp] = {
+    val timer = metrics.defaultRegistry.timer("ppt.addresslookup.initialise.timer").time()
+    http.POST[AddressLookupConfigV2, AddressLookupOnRamp](appConfig.addressLookupInitUrl,
+                                                          addressLookupRequest
+    )(implicitly, InitialiseAddressLookupReads, hc, ec)
+      .andThen { case _ => timer.stop() }
+  }
+
+  private[connectors] def getAddressUrl(id: String) =
```
[the remark this answers is not in the document]
Comment (SteveSugden) [reply app/uk/gov/hmrc/plasticpackagingtax/registration/connectors/addresslookup/AddressLookupFrontendConnector.scala:52]: Can probably just be "private" to this class

File: app/uk/gov/hmrc/plasticpackagingtax/registration/controllers/contact/ContactDetailsAddressController.scala [hunk 4]
```diff
@@ -78,6 +92,34 @@ class ContactDetailsAddressController @Inject() (
...
+    }
+
+  def update(id: Option[String]): Action[AnyContent] =
+    (authenticate andThen journeyAction).async { implicit request =>
+      addressLookupFrontendConnector.getAddress(
+        id.getOrElse(throw new MissingAddressIdException)
+      ).flatMap {
+        confirmedAddress =>
+          val updatedAddress = Address(confirmedAddress)
+          // Address Lookup Service may return an address that is incompatible with PPT, so validate it again
```
[the remark this answers is not in the document]
Comment (SteveSugden) [reply app/uk/gov/hmrc/plasticpackagingtax/registration/controllers/contact/ContactDetailsAddressController.scala:110]: No, we'll still need this.  ALF is not guaranteed to return a postcode (even in uk-mode) if the user uses the manual entry method

File: app/uk/gov/hmrc/plasticpackagingtax/registration/models/addresslookup/AddressLookupConfigV2.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (SteveSugden) [reply app/uk/gov/hmrc/plasticpackagingtax/registration/models/addresslookup/AddressLookupConfigV2.scala:36]: We do look up messages via a key.  Currently only this one - 
```
AppLevelLabels(navTitle = en("service.name"))
```

We need to send over the message values (not the keys) to ALF

Welsh would need to be added later.  To do this you need to have a second  set of "LanguageLabels".  Something like this 
```
                              labels = JourneyLabels(en =
                                LanguageLabels(appLevelLabels =
                                  AppLevelLabels(navTitle = en("service.name"))
                                ),
                                cy =
                                  LanguageLabels(appLevelLabels =
                                    AppLevelLabels(navTitle = cy("service.name"))
                                  )
                              )
```

File: app/uk/gov/hmrc/plasticpackagingtax/registration/models/addresslookup/AddressLookupConfirmation.scala (hunk 2 above)
Comment (SteveSugden) [reply app/uk/gov/hmrc/plasticpackagingtax/registration/models/addresslookup/AddressLookupConfirmation.scala:35]: Agreed.  I might refactor this method based on some similar code I'm writing for the back-end

File: app/uk/gov/hmrc/plasticpackagingtax/registration/connectors/addresslookup/AddressLookupFrontendConnector.scala [hunk 5]
```diff
@@ -43,38 +42,24 @@ class AddressLookupFrontendConnector @Inject() (
...
+    http.POST[AddressLookupConfigV2, HttpResponse](appConfig.addressLookupInitUrl,
+                                                   addressLookupRequest
+    ).andThen { case _ => timer.stop() }
+      .map {
+        case response @ HttpResponse(ACCEPTED, _, _) =>
+          response.header(LOCATION) match {
+            case Some(redirectUrl) => AddressLookupOnRamp(redirectUrl)
+            case None              => throw new IllegalStateException("Missing re-direct url")
+          }
+      }
```
Comment (RyanKellyHMRC) [inline app/uk/gov/hmrc/plasticpackagingtax/registration/connectors/addresslookup/AddressLookupFrontendConnector.scala:54]: 🙇

File: app/uk/gov/hmrc/plasticpackagingtax/registration/models/addresslookup/AddressLookupConfigV2.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (RyanKellyHMRC) [reply app/uk/gov/hmrc/plasticpackagingtax/registration/models/addresslookup/AddressLookupConfigV2.scala:34]: Thanks for the explanation. ⭐

**Review by RyanKellyHMRC — APPROVED** [review]
TY ⭐

**Pull request merged**